### PR TITLE
volume analysis

### LINKED ISSUE #200: unify /volume and /event-aggregates into /analytics
merge the /volume functionality into /event-aggregates, however:

* supports both cases: either authenticate as someone and get their stats only, or get global stats; however, global stats would be limited to certain combinations of periods and intervals (e.g. daily for a month, hourly for a day, etc.)
* the authenticated calls should get data either for a single channel, or for all channels; the global calls will always be for all channels
* rather than using the $day, $month and etc. mongodb operators, use what we use in volume (`$mod`): https://github.com/AdExNetwork/adex-validator/blob/05-stable/routes/volume.js#L29 ; each datapoint will use `time` as it's identifier, and should return it in a number format (rather than stringified date) ; this should greatly simplify the pipeline
* to avoid excessive CPU load, calls without authentication, that retrieve global stats, should be cached in redis for interval/2 time
* should be all handled in a separate file, `routes/analytics.js`; might use the channels middleware if we need to load a full channel

for example, `aggr` would be `[{ time: 1563645559247, value: "1023022" }]`

API:

* `/analytics` - optional authentication; queries for all channels
* `/:channelId/analytics` - required authentication; channel specific queries

If authentication is passed, it will only return aggregates for this particular earner; otherwise it will return a sum of all earners (like `/volume` now)

Both take `timeframe`, `limit`, `eventType`, `metric`; only certain combinations of `timeframe` and `limit` will be allowed for `/analytics` without authentication

**Since this is a breaking change, it should be discussed with @IvoPaunov and @samparsky first**

### PR DESCRIPTION
**Pull request by samparsky**
Fix #200

### CODE REVIEW

File: routes/analytics.js [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+
+	const pipeline = [
+		{ $match: { ...match } },
+		{
+			$project: {
+				created: 1,
+				value: {
+					$sum: {
+						$map: {
+							input: { $objectToArray: `$events.IMPRESSION.eventPayouts` },
```
Comment (Ivshti) [inline routes/analytics.js]: this should work on counts too; it should be a parameter like event-aggregates

the event type (IMPRESSION) should be a parameter too

File: routes/analytics.js [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+	if (req.channel) {
+		match = { ...match, channelId: req.channel.id }
+	}
+	if (period) {
+		const after = +req.query.after || 0
+		match = { ...match, created: { $gt: new Date(new Date(after).getTime() - period) } }
+	}
+
+	const pipeline = [
+		{ $match: { ...match } },
```
Comment (Ivshti) [inline routes/analytics.js]: why not just `$match: match`

File: routes/analytics.js [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+		},
+		{
+			$group: {
+				_id: {
+					$subtract: [{ $toLong: '$created' }, { $mod: [{ $toLong: '$created' }, interval] }]
+				},
+				value: { $sum: '$value' }
+			}
+		},
+		{ $sort: { _id: 1 } },
```
Comment (Ivshti) [inline routes/analytics.js]: we should think about the `match` and the MongoDB indexes, cause if the data comes unordered in here, we might run out of memory while sorting:

```
MongoError: Sort exceeded memory limit of 104857600 bytes, but did not opt in to external sorting. Aborting operation. Pass allowDiskUse:true to opt in.
    at queryCallback (/app/node_modules/mongodb-core/lib/cursor.js:248:25)
    at /app/node_modules/mongodb-core/lib/connection/pool.js:532:18
    at process._tickCallback (internal/process/next_tick.js:61:11)

```

Comment (Ivshti) [reply routes/analytics.js]: we have an index of `{ channelId: 1, created: 1}` which would be perfect here

perhaps we should hint the pipeline to use this index?

Comment (samparsky) [reply routes/analytics.js]: I think we should add both the index and `{allowDiskUse: true}`

**Comment by Ivshti** [comment]
@samparsky I will approve/merge that when the relevant PR is made in adex-explorer

**Comment by Ivshti** [comment]
bugs/issues:

* setting period to 0 for authenticated users doesn't make any sense cause you allow `after` to be set only if `period` is truthy
* `allowDiskUse` shouldn't be needed if indexes are done right; using it is not a good practice
* never convert to a number with `+`, it's not intuitive
* getting user-specific data doesn't really work; it only **matches** aggregates that contain that user, but it still sums all the stats from them, rather than only extracting the user-specific stats
* `eventCounts` now saves `${uid}:${adUnitId}`, so not even the `$match` for that would work, since it's a strict match for `eventCounts.${uid}`
* caching does not account for the UID, so one user would be getting the cached data for the user who previously queried

don't touch for now, I'm rewriting

**Comment by Ivshti** [comment]
another one

* must be `celebrate({ query: ...` rather than `body`

**Pull request merged**